### PR TITLE
WindowServer+LibGUI+a bunch others: Isolate WindowManager, add winlist and winmove utilities

### PR DESCRIPTION
**Pull request by demostanis**
I wanted to make a small command to move windows, `winmove`, and realized that the current WindowManager API required a Window, although that should not really be needed. Thus, I isolated it, and made the changes in Taskbar, Keymap applet and WorkspacePicker applet to use the new way of instantiating a WindowManager.
I also added a command to list windows, `winlist`, to know the ID of the one to move in `winmove`, which uses that new method, and makes use of a new GreetingIsOver event, which allows it to exit when WindowServer has stopped greeting it (sent the current window list), or to stay alive and watch for new windows (with `-k` switch, and it actually watches for window changes like focus, size changes, might change this behavior in the future).
Man pages included.

### CODE REVIEW

File: Userland/Libraries/LibGUI/Window.cpp [hunk 1]
```diff
@@ -321,14 +321,6 @@ void Window::set_window_type(WindowType window_type)
...
 }
 
-void Window::make_window_manager(unsigned event_mask)
-{
-    (void)event_mask;
-    // FIXME: delete this function
-    // GUI::ConnectionToWindowManagerServer::the().async_set_event_mask(event_mask);
-    // GUI::ConnectionToWindowManagerServer::the().async_set_manager_window(m_window_id);
-}
-
```
Comment (networkException) [inline Userland/Libraries/LibGUI/Window.cpp:329]: Removing the function here seems arbitrary. It doesn't really fit in the scope of the commit and is not clear from its message. Also trying to compile serenity after this would fail as Keymap still relies on the function

File: Userland/Utilities/winlist.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2022, demostanis worlds <demostanis@protonmail.com>
+ * Copyright (c) 2022, the SerenityOS developers.
```
Comment (networkException) [inline Userland/Utilities/winlist.cpp:3]: Why two copyright holders?

File: Userland/Utilities/winmove.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2022, demostanis worlds <demostanis@protonmail.com>
+ * Copyright (c) 2022, the SerenityOS developers.
```
Comment (networkException) [inline Userland/Utilities/winmove.cpp:3]: Same here. Everywhere else you use "the SerenityOS developers." or "The SerenityOS developers", I would suggest being consistent about this

**Review by networkException — COMMENTED** [review]
Overall the commits could use descriptions about their intention

File: Userland/Utilities/winlist.cpp (hunk 2 above)
Comment (demostanis) [reply Userland/Utilities/winlist.cpp:3]: I just wanted to have my name somewhere and saw some other files do it. Guess I'll keep "the SerenityOS developers"

File: Userland/Libraries/LibGUI/Window.cpp (hunk 1 above)
Comment (demostanis) [reply Userland/Libraries/LibGUI/Window.cpp]: my bad

File: Userland/Utilities/winlist.cpp (hunk 2 above)
Comment (networkException) [reply Userland/Utilities/winlist.cpp]: I didn't mean to discourage you to put your name on the changes, just adding two copyright headers refering to the same author for this change is weird.

If you use `demostanis worlds <demostanis@protonmail.com>` then its probably better to do so for all the files

**Review by networkException — APPROVED** [review]
Code wise LGTM

I'm not a huge fan of having `winlist` and `winmove` be separate utilities (although thats probably more unix-y). I also think `winlist` should default to the equivalent of `winlist -ti` for the default output because its quite useless otherwise. And lastly `winmove` should probably account for window frames (or at least have an option for that, I'll probably implement it myself if you don't want to) - all of this is subjective though and a maintainer might think otherwise

**Comment by demostanis** [comment]
ahem what have i done with the commit history lol

**Comment by demostanis** [comment]
I am in favor of defaulting to `winlist -it`, and I changed the PR to reflect it. As for having `win*` commands as one, I don't think that's necessary, it's better to separate them as they don't really do the same thing. 
I'll let you do the taking-into-account-window-frames part :) :yaklazy:

File: Userland/Libraries/LibGUI/WindowManager.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace GUI {
+
+class WindowManager : public Core::Object {
+    C_OBJECT(WindowManager)
+
+public:
+    static WindowManager* from_wm_id(int);
+    virtual ~WindowManager() override;
+    int wm_id() { return m_wm_id; };
```
Comment (timschumi) [inline Userland/Libraries/LibGUI/WindowManager.h:20]: "Meh" on introducing new heavily abbreviated function and variable names to the code base. Those should be `window_manager_id` and `m_window_manager_id` respectively (same for other occurrences of `wm`).

File: Userland/Libraries/LibGUI/Window.cpp [hunk 5]
```diff
@@ -326,8 +326,10 @@ void Window::set_window_mode(WindowMode mode)
 
 void Window::make_window_manager(unsigned event_mask)
 {
-    GUI::ConnectionToWindowManagerServer::the().async_set_event_mask(event_mask);
-    GUI::ConnectionToWindowManagerServer::the().async_set_manager_window(m_window_id);
+    (void)event_mask;
+    // FIXME: delete this function
+    // GUI::ConnectionToWindowManagerServer::the().async_set_event_mask(event_mask);
+    // GUI::ConnectionToWindowManagerServer::the().async_set_manager_window(m_window_id);
```
Comment (timschumi) [inline Userland/Libraries/LibGUI/Window.cpp]: This function should be refactored away first if possible.

File: Userland/Services/WindowServer/ConnectionFromClient.cpp [hunk 6]
```diff
@@ -434,7 +434,7 @@ Messages::WindowServer::SetWindowRectResponse ConnectionFromClient::set_window_r
 {
     auto it = m_windows.find(window_id);
     if (it == m_windows.end()) {
-        did_misbehave("SetWindowRect: Bad window ID");
+        did_misbehave(String::formatted("SetWindowRect: Bad window ID: {}", window_id).characters());
```
Comment (timschumi) [inline Userland/Services/WindowServer/ConnectionFromClient.cpp:437]: This looks like a change that is not related to setting window rects.

File: Userland/Services/WindowServer/WMConnectionFromClient.cpp [hunk 7]
```diff
@@ -23,7 +24,8 @@ WMConnectionFromClient::~WMConnectionFromClient()
 {
     // The WM has gone away, so take away the applet manager (cause there's nowhere
     // to draw it in).
-    AppletManager::the().set_position({});
+    if (m_manages_applets)
+        AppletManager::the().set_position({});
```
Comment (timschumi) [inline Userland/Services/WindowServer/WMConnectionFromClient.cpp:28]: This also looks unrelated to setting rects.

File: Userland/Services/WindowServer/WindowManager.cpp [hunk 8]
```diff
@@ -419,6 +419,14 @@ void WindowManager::greet_window_manager(WMConnectionFromClient& conn)
...
     if (auto* applet_area_window = AppletManager::the().window())
         tell_wms_applet_area_size_changed(applet_area_window->size());
+    tell_wm_greeting_is_over(conn);
+}
+
+void WindowManager::tell_wm_greeting_is_over(WMConnectionFromClient& conn)
+{
+    if (conn.wm_id() < 0)
+        return;
+    conn.async_greeting_is_over(conn.wm_id());
```
Comment (timschumi) [inline Userland/Services/WindowServer/WindowManager.cpp:429]: This isn't really "dead code", but it looks like it doesn't serve a purpose yet.

File: Userland/Applets/Keymap/main.cpp [hunk 9]
```diff
@@ -18,11 +19,15 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
+    KeymapWindowManager wm(window);
     window->set_has_alpha_channel(true);
     window->set_title("Keymap");
     window->resize(16, 16);
     window->show();
-    window->make_window_manager(WindowServer::WMEventMask::KeymapChanged);
+
+    GUI::ConnectionToWindowManagerServer::the().async_set_event_mask(
+        WindowServer::WMEventMask::KeymapChanged);
+    GUI::ConnectionToWindowManagerServer::the().async_set_window_manager(wm.wm_id(), true);
```
Comment (timschumi) [inline Userland/Applets/Keymap/main.cpp:30]: Again, this refactor should be done before doing the WindowManager split. The rest of the changes should then be squashed into the split itself.

File: Userland/Applets/WorkspacePicker/main.cpp [hunk 10]
```diff
@@ -42,7 +45,9 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     auto show_tray_icon = [&] {
         if (!window->is_visible()) {
             window->show();
-            window->make_window_manager(WindowServer::WMEventMask::WorkspaceChanges);
+            GUI::ConnectionToWindowManagerServer::the().async_set_event_mask(
+                WindowServer::WMEventMask::WorkspaceChanges);
+            GUI::ConnectionToWindowManagerServer::the().async_set_window_manager(wm.wm_id(), true);
         }
```
Comment (timschumi) [inline Userland/Applets/WorkspacePicker/main.cpp:51]: Same here.

File: Userland/Services/Taskbar/main.cpp [hunk 11]
```diff
@@ -81,7 +81,7 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
         | WindowServer::WMEventMask::WindowRemovals
         | WindowServer::WMEventMask::WindowIconChanges
         | WindowServer::WMEventMask::WorkspaceChanges);
-    GUI::ConnectionToWindowManagerServer::the().async_set_window_manager(wm.wm_id());
+    GUI::ConnectionToWindowManagerServer::the().async_set_window_manager(wm.wm_id(), true);
```
Comment (timschumi) [inline Userland/Services/Taskbar/main.cpp:84]: Looks like this is a fixup to one of the previous commits?

File: Base/usr/share/man/man1/winlist.md [hunk 12]
```diff
@@ -0,0 +1,51 @@
+## Name
+
+winlist - List windows
```
Comment (timschumi) [inline Base/usr/share/man/man1/winlist.md:3]: I personally feel like the manpage additions should probably be squashed into the commit that adds the utility itself, but others may have other opinions on that.

**Review by timschumi — CHANGES_REQUESTED** [review]

File: Userland/Services/WindowServer/WindowManager.cpp (hunk 8 above)
Comment (demostanis) [reply Userland/Services/WindowServer/WindowManager.cpp:429]: It does, that's used in winlist

Comment (timschumi) [reply Userland/Services/WindowServer/WindowManager.cpp:429]: I mean, yeah, in that sense it is related, but not to the commit that adds the rect setting API itself, especially since the other half of the `greeting_is_over` handling only gets introduced in a later commit.

File: Userland/Services/WindowServer/ConnectionFromClient.cpp (hunk 6 above)
Comment (demostanis) [reply Userland/Services/WindowServer/ConnectionFromClient.cpp:437]: right, that's a debugging leftover...

File: Userland/Libraries/LibGUI/WindowManager.h (hunk 4 above)
Comment (demostanis) [reply Userland/Libraries/LibGUI/WindowManager.h:20]: i didn't introduce it, wm and wm_id was used before i did
as for using id() instead of wm_id, Window.cpp uses window_id(), so i decided to do something alike

Comment (demostanis) [reply Userland/Libraries/LibGUI/WindowManager.h:20]: (it's also much easier to search for wm_id than id)

File: Userland/Services/WindowServer/WindowManager.cpp (hunk 8 above)
Comment (demostanis) [reply Userland/Services/WindowServer/WindowManager.cpp:429]: no idea how it ended up in that commit
i wonder if i will ever be good at using git someday 😩

File: Userland/Libraries/LibGUI/WindowManager.h (hunk 4 above)
Comment (timschumi) [reply Userland/Libraries/LibGUI/WindowManager.h:20]: `id` would only be used internally though, or when using a WindowManager object (i.e. `window_manager.id()`). Everything else should be named `window_manager_id` instead.

Since we are touching most of the call sites anyways, I don't think there is much harm in giving it a proper name now.

Comment (demostanis) [reply Userland/Libraries/LibGUI/WindowManager.h:20]: i wonder if that's really useful, we already know the context in every place where wm_id (or wm) is used, making it clear wm stands for window manager

File: Userland/Applets/Keymap/main.cpp [hunk 13]
```diff
@@ -22,7 +22,9 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     window->set_title("Keymap");
```
Comment (timschumi) [inline Userland/Applets/Keymap/main.cpp:24]: Can't comment on the commit message directly, but the title category here is probably fine with just `LibGUI:`. All the others are just users of that API, and mixing four different categories for that message seems a bit crowded.

File: Userland/Applets/Keymap/main.cpp [hunk 14]
```diff
@@ -22,7 +22,9 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
     window->resize(16, 16);
     window->show();
-    window->make_window_manager(WindowServer::WMEventMask::KeymapChanged);
+
+    GUI::ConnectionToWindowManagerServer::the().async_set_event_mask(WindowServer::WMEventMask::KeymapChanged);
```
Comment (timschumi) [inline Userland/Applets/Keymap/main.cpp]: This line gets reformatted in the next commit (without any changes as far as I can tell), same with some of the other occurrences of `async_set_event_mask`.

Please run through a rebase again, and make sure to run the formatter and commit the changes that it makes after this commit and after the following one. This ensures that the formatting stays consistent.

**Review by timschumi — CHANGES_REQUESTED** [review]
It already looks much nicer. :^)

One last round of feedback from me. After that, someone who actually knows LibGUI probably needs to look at this instead of me.

File: Userland/Services/Taskbar/main.cpp [hunk 15]
```diff
@@ -74,12 +75,8 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
     window->show();
 
-    GUI::ConnectionToWindowManagerServer::the().async_set_event_mask(
-        WindowServer::WMEventMask::WindowStateChanges
-        | WindowServer::WMEventMask::WindowRemovals
-        | WindowServer::WMEventMask::WindowIconChanges
-        | WindowServer::WMEventMask::WorkspaceChanges);
-    GUI::ConnectionToWindowManagerServer::the().async_set_window_manager(window.window_id());
+    TaskbarWindowManager wm(window);
+    GUI::ConnectionToWindowManagerServer::the().async_set_window_manager(wm.wm_id());
```
[the remark this answers is not in the document]
Comment (demostanis) [reply Userland/Services/Taskbar/main.cpp]: it went away along with my cautiousness

File: Userland/Applets/Keymap/main.cpp (hunk 14 above)
Comment (demostanis) [reply Userland/Applets/Keymap/main.cpp]: actually it seemed like clang-format did not do anything to that line, it would accept both

**Comment by demostanis** [comment]
i hope i didn't mess up with git

File: Userland/Services/WindowServer/WMConnectionFromClient.cpp [hunk 16]
```diff
@@ -186,5 +186,4 @@ void WMConnectionFromClient::set_keymap(String const& keymap)
 {
     WindowManager::the().keymap_switcher()->set_keymap(keymap);
 }
-
```
Comment (timschumi) [inline Userland/Services/WindowServer/WMConnectionFromClient.cpp:189]: This is a (non style-conforming) change that gets reverted in a later commit, so that should not happen at all.

**Review by timschumi — CHANGES_REQUESTED** [review]
The smallest of nitpicks that you have probably ever seen.

**Comment by demostanis** [comment]
Reading back the code, it actually seems pretty weird. The current methods (without this PR) work on the WindowManager's managed windows (with the exception of show_desktop()), while I added methods which work on all windows (to make winmove and winlist possible).
Should I make all methods in ConnectionFromClient work on all windows? Is the current state good? Should I add another API instead?
